### PR TITLE
Documentation improvement

### PR DESCRIPTION
**Pull request by AntoineRondelet**
This branch contains a README which resumes the list of commands a user has to run in order to do a payment through Mobius.
Please try and run the "tutorial" and tell me if something's missing/unclear :)

### CODE REVIEW

File: mobiusTutorial.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    "extraData"  : "",
+    "gasLimit"   : "0xFFFFFFF",
+    "gasPrice"   : "0x1",
+    "nonce"      : "0x0000000000000042",
+    "mixhash"    : "0x0000000000000000000000000000000000000000000000000000000000000000",
+    "parentHash" : "0x0000000000000000000000000000000000000000000000000000000000000000",
+    "timestamp"  : "0x00"
+}
+```
+*Note:* Make sure to have the line `"byzantiumBlock": 0` in the `config` of your `genesis.json` file in order tomake sure that the Mobius contract will be able to execute precompiled contracts. (See: [Byzantium](https://blog.ethereum.org/2017/10/12/byzantium-hf-announcement/))
```
Comment (shapeshed) [inline mobiusTutorial.md]: tomake -> to make

**Comment by AntoineRondelet** [comment]
Thanks for the typo @shapeshed, and of course, the abi shouldn't have been committed, I explained how to generate it in the tuto, so it was an error.
Should be fixed now :)

File: mobiusTutorial.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+```bash
+$ node --version
+v9.5.0
+```
+
+## How to use Mobius ?
+
+### Assumptions
+
+1. The `Mixer` has a `RING_SIZE` equal to 1 (see: https://github.com/clearmatics/mobius/blob/master/contracts/LinkableRing.sol#L80).
```
Comment (HarryR) [inline mobiusTutorial.md]: Ring size is 4.

**Comment by HarryR** [comment]
IMO the ABI should be committed. It allows people to more easily use a deployed contract without having to go through the multi-stage compile process.

If it becomes stale you have bigger problems, just add it into the Makefile.

**Comment by shapeshed** [comment]
I agree promoting ease of use is important so don't have any objections to the ABI being included. Providing the path of least resistance is obviously important at this phase of the project.

File: mobiusTutorial.md (hunk 2 above)
Comment (AntoineRondelet) [reply mobiusTutorial.md]: This is an hardcoded value. I assume that the ring size is 1 to make the tuto easier and shorter. The file has to be edited before the contracts are deployed. See:
```
Assumptions

    The Mixer has a RING_SIZE equal to 1 (see: https://github.com/clearmatics/mobius/blob/master/contracts/LinkableRing.sol#L80). If you want to follow the rest of the tutorial, please clone this repository and modify the LinkableRing contract accordingly:

export $WORKDIR = ~/Path/To/Your/Working/Directory
git clone https://github.com/clearmatics/mobius.git $WORKDIR
cd $WORKDIR/mobius
[YourFavoriteTextEditor] contracts/LinkableRing.sol
[Change line 80: uint256 public constant RING_SIZE = 4; INTO uint256 public constant RING_SIZE = 1;]
[Save and exit your editor]
```
The purpose here is not to simulate a real flow but just to show how things work together. Then people are free to change the RING_SIZE as they want to play around.

**Comment by HarryR** [comment]
To reduce the 'path of least resistance' you need a CLI app which you can point at an EthRPC instance and let it do its thing.

**Comment by AntoineRondelet** [comment]
Hum... What would the `mix` command do @HarryR ? I don't really get the idea here. Please tell me if I am wrong, but as far as I understand, the privacy solution proposed by Mobius relies exclusively on the use of ring signatures along with stealth key pairs. Thus, we are not really doing any mixing, but more transfer obfuscation. Thus if Alice wants to send Bob 1 ether, she just has to send 1 ether to the Mixer along with Bob's stealth public key (derived from Bob's master public key), in order to constitute the ring. In order to withdraw Bob has to wait for the ring to be full and derives some temporary keys in order to generate a new address and sign to take his funds from the Mixer. 
Looking at the following scenario, I don't really understand what you mean by 'mixing operations' ? As far as I understand the funds that Alice sends to the Mixer go to Bob (and not to another recipient, like in "normal mixing"), the properties of stealth addresses and ring sig just make inference of the recipient super hard.

For the CLI, I agree that it would be a good idea. I'll look into it, however, this would be the object of another PR. This PR only aims to recap the manipulations that one has to play around with Mobius as it is implemented for now :)

**Comment by HarryR** [comment]
Yup I totally agree this would be a separate issue, but I'm just getting my thoughts out there and trying to think of ways that would make it more usable to a wider audience.

So the problem with mixing, and the `mix` command, is that I don't think one round is not enough for sufficient privacy. My reasoning behind it is as follows:

 * With a ring size of 4 there's a 25% probability of any input being related to any output, if I deposit 1 ETH and you withdraw 1 ETH, any observer will know that the two parties are related with a probability of 1/4.

 * If you try and send an amount, broken into denominations, e.g. 1 ETH, 0.2 ETH and 0.03 ETH (sums to 1.23 ETH), the same Ethereum address will deposit all 3, and a different Ethereum address will withdraw all 3. At this point neither side has any anonymity, I can see 1.23 deposited, 1.23 withdrawn, therefore I know Alice send 1.23 to Bob.

 * Stealth addresses and one-time keys don't matter if the gas payer (e.g. Alice's or Bob's account) links multiple deposits or withdraws to each other. I am looking at the weakest links, pieces of information which allow transactions to be correlated or linked to each other etc. Such as: total or specific amounts, gas payer, multiple rings with the same participants etc.

Inference of the sender and recipient isn't that difficult, CoinJoin/CoinShuffle is fundamentally broken from an anonymity perspective, as are most BitCoin mixers (like... in-practice, people gone to jail because of it, it simply doesn't offer strong enough privacy guarantees). With Ethereum you are at even more of a disadvantage because the account which pays for Gas links all the transactions together.

>  In order to withdraw Bob has to wait for the ring to be full and derives some temporary keys in order to generate a new address and sign to take his funds from the Mixer.

Then Bob submits the ring signature from his account, or puts enough Gas into a new account and uses that to submit the ring signature to withdraw etc.

Lets think about side-channels for a minute, and you'll see what I mean:

  * Gas paying account, or whichever account deposited the ETH into the account, aka the originator of the deposit/withdraw transaction
  * Denomination splitting, if deposits and withdraws of a total sum amount, even from multiple rings, move at the same time then they are correlated with near 100% accuracy and privacy is reduced to 0%...
  * 25% probability of a single transaction being linked is pretty low, anything which can link multiple transactions together increases that exponentially closer to 100%...
  * Any ring you participate in every other person has a 1/3 chance of being linked by you, at relatively little cost I can be the other 3 participants in all of the rings, by 'filling' them your transaction goes through quicker - but you also lose 100% anonymity. The 'wait for the ring to fill' is a big problem because it allows anybody else to de-anonymise you by actively participating.

So, how do we tackle this?

 1) By carefully and repeatedly performing 'mix' operations you are reducing the probability of Alice and Bob being linked. A mix operation is depositing into another ring directly from a ring without the funds being temporarily deposited. After 1 mix operation the probability of Alice and Bob being linked should around 6% (`1 / 4 / 4` ? or am I missing some probability theory there...), the more mixing operations you perform the lower the probability gets until the point where everybody is statistically related to everybody else (the ideal). For `mix`, instead of `Withdraw` it would be `WithdrawThenDeposit` - you provide it with the ring signature for `Withdraw` and the public key to be used when it's deposited into the next ring.

 2) Provide a proxy service which will anonymously submit transactions on your behalf to obscure the origin address, if you perform this over Tor, or with some kind of insecure legal BS contract which says 'I promise to not deanonymise you' (which... is not a strong privacy guarantee) then correlating Alice and Bob becomes more difficult as long as you cannot correlate the two by seeing the same denominations from the same rings move at the same time.

  3) Wait for 'account abstraction' which means the mixing contract can pay for its own gas, e.g. you send from '000...' address and the contract pays the miner (but oh, wait... in 'enterprise blockchain' there are no miners, or Gas... so wtf am I even rambling on about). But you still have the problem of the total sum of the denominations moving at the same time making it easy to link deposit and withdraw unless it's carefully spread out over time.

  4) Don't do denominational splitting, Monero tried that and it got deanonymised very quickly... this isn't a hypothetical academic attack, this is real. DO NOT DO DENOMINATIONAL SPLITTING UNLESS YOU ARE VERY VERY CAREFUL ABOUT SPREADING THE INFORMATION OUT OVER A LONG TIME PERIOD.

  5) Use RingCT

  6) Use Monero.

**Comment by AntoineRondelet** [comment]
The weak point of Mobius is indeed the fact that the recipient has to pay some gas to retrieve his funds. This is undoubtedly the door wide open for an attacker to carry out an analysis and infer link the recipient's multiple addresses.

>  If you try and send an amount, broken into denominations, e.g. 1 ETH, 0.2 ETH and 0.03 ETH (sums to 1.23 ETH), the same Ethereum address will deposit all 3, and a different Ethereum addresses will withdraw all 3. At this point neither side has any anonymity, I can see 1.23 deposited, 1.23 withdrawn, therefore I know Alice send 1.23 to Bob.

I'm not sure if this is as straightforward as you described it. In the case where Alice wants to send 1.23 eth to Bob, split into 1, 0.2 and 0.03 denominations, she needs to generate 3 stealth public keys from Bob's master public key. Here the "stealthiness" protects Bob. A passive attacker would not be able to distinguish between the case where Alice is doing a split payment to Bob (unique recipient) OR the case where she is doing a payment to 2 or 3 different recipients. Thus Bob should be able retrieve the funds without being worried about his identity being leaked by Alice's deposit. However, once again, Bob needs to have gas on his accounts to pay the gas of the withdrawals... which is the point where the attacker would be able to exploit side-channel leaks of information to threat Bob's anonymity. 

> The 'wait for the ring to fill' is a big problem because it allows anybody else to de-anonymize you by actively participating.

Interesting attack scenario indeed ! The n-1 participants of a ring of size n could collude to de-anonymize a participant (Bob) by filling the ring.
I do not really see how using "multiple layers" of rings, alone, could protect Bob in such a scenario. The only solution that appears to me (although I would need to put more thoughts into it) would be to use multiple layers (or mixing rounds) of rings of *DIFFERENT SIZES*. That way the n-1 attackers' payments trying to unveil the identity of Bob would be spread out over multiple rings and other anonymous recipients. This would spread the group of attackers and introduce indistinguishably during the withdrawal.

**Comment by HarryR** [comment]
> I'm not sure if this is as straightforward as you described it. In the case where Alice wants to send 1.23 eth to Bob, split into 1, 0.2 and 0.03 denominations, she needs to generate 3 stealth public keys from Bob's master public key.

Ignore stealth keys, they are noise.  Tag X/Y is noise. Ignore the noise.

| Who | Action | Amount |
| ---- | ----- | ---- |
| Alice | Deposit | 1.0 |  
| Alice | Deposit | 0.2 |
| Alice | Deposit | 0.03 |
| Bob | Withdraw | 1.0 |
| Bob | Withdraw | 0.2 |
| Bob | Withdraw | 0.03 |

**Comment by HarryR** [comment]
Lets work on something more productive instead: https://github.com/clearmatics/mobius/issues/12

**Comment by HarryR** [comment]
Also, see: http://coinjoinsudoku.com/

**Comment by AntoineRondelet** [comment]
> I have personally seen this approach being used to de-anonymise IRL, but without the 'Who' column...

In the case of Mobius, your array would most likely look like:

Who | Action | Amount
-- | -- | --
Alice (using Bob's spk1) | Deposit | 1.0
Alice (using Bob's spk2) | Deposit | 0.2
Alice (using Bob's spk3) | Deposit | 0.03
Bob (address 1) | Withdraw | 1.0
Bob (address 2) | Withdraw | 0.2
Bob (address 3) | Withdraw | 0.03

Which would translate in something like this from an attacker perspective:

Who | Action | Amount
-- | -- | --
Alice | Deposit | 1.0
Alice | Deposit | 0.2
Alice | Deposit | 0.03
Recipient 1 | Withdraw | 1.0
Recipient 2 | Withdraw | 0.2
Recipient 3 | Withdraw | 0.03

Now regardless of the gas problem for the recipient, the attacker here has to answer the question: "What are the links between recipient1, 2 and 3 ?", to which the answers would either be:
1. Recipient 1 = recipient2 = recipient3
2. Two recipients are the same and one is different
3. All recipients are different.

Answering this question with a decent probability is pretty hard in the case where we assume that:
1) The problem of the gas payer is ignored (each newly created address has a sufficient balance to call the Withdraw function)
2) A sufficient amount of transactions are handled by the system (A large amount of participants doing a large amount of Tx on the network). This hypothesis is important to guarantee that all 3 scenarios of the question `"What are the links between recipient1, 2 and 3 ?"` are indistinguishable.

My question here is: "How an attacker could point determine the relationship between recipient1,2 and 3 with good probability ?". As first glance, timing information doesn't seem to be sufficient to guarantee the success of the attack (thanks to hypothesis 2). Moreover, `Amount` could be used to link Alice to one of the recipient, but the relationship between recipients is still obfuscated in this case...

Anyway, the answer to my question might be presented in academic publications that I haven't read yet. Do you have good resources to provide me on these points ?

**Comment by HarryR** [comment]
See:

 * http://coinjoinsudoku.com/
 * https://arxiv.org/pdf/1712.01210.pdf
 * https://medium.com/@nopara73/tumblebit-vs-coinjoin-15e5a7d58e3

I give up, enjoy the project m8.

**Pull request merged**